File: sandbox/src/validation.c
/* Host-side input validation implementation. */
#include "validation.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/stat.h>
#ifdef _WIN32
#include <windows.h>
#include <io.h>
#define access _access
#ifndef R_OK
#define R_OK 04
#endif
#else
#include <unistd.h>
#include <limits.h>
#endif

int validate_input(const char *path, char *error, size_t error_size) {
    struct stat st;
    char resolved[4096];
    const char *p;

    if (!path || !error || error_size == 0) return -1;

    /* Resolve real path */
#ifdef _WIN32
    if (!_fullpath(resolved, path, sizeof(resolved))) {
        snprintf(error, error_size, "Path resolution failed");
        return -1;
    }
#else
    if (!realpath(path, resolved)) {
        snprintf(error, error_size, "Path resolution failed: %s", path);
        return -1;
    }
#endif

    /* Exists? */
    if (stat(resolved, &st) != 0) {
        snprintf(error, error_size, "File not found: %s", resolved);
        return -1;
    }

    /* Is regular file? */
#ifdef _WIN32
    if (!(st.st_mode & _S_IFREG)) {
#else
    if (!S_ISREG(st.st_mode)) {
#endif
        snprintf(error, error_size, "Path is not a regular file");
        return -1;
    }

    /* Size cap */
    if (st.st_size > (off_t)MAX_FILE_SIZE_BYTES) {
        double actual_mb = (double)st.st_size / (1024.0 * 1024.0);
        snprintf(error, error_size,
                 "File too large: %.1f MB exceeds maximum allowed size of 500 MB",
                 actual_mb);
        return -1;
    }

    if (st.st_size == 0) {
        snprintf(error, error_size, "File is empty (0 bytes)");
        return -1;
    }

    /* Read permission */
    if (access(resolved, R_OK) != 0) {
        snprintf(error, error_size, "Permission denied: cannot read file");
        return -1;
    }

    /* Path traversal check */
    p = resolved;
    while (*p) {
        if (p[0] == '.' && p[1] == '.' &&
            (p == resolved || *(p-1) == '/' || *(p-1) == '\\')) {
            /* Only flag obvious traversal patterns */
            /* We use realpath above which already resolves these */
            break;
        }
        p++;
    }

    return 0;
}

File: sandbox/validation.py
"""Host-side input validation for the sandbox file analysis engine.

Validates file existence, size caps, read permissions, and prevents
path traversal attacks before any security layer is invoked.
"""

import os
import stat

from config import MAX_FILE_SIZE_BYTES


def validate_input(file_path: str) -> dict | None:
    """Validate the input file before pipeline execution.

    Returns None if the file passes all checks.
    Returns an error dict (suitable for direct emit) if validation fails.
    """
    # --- Resolve to absolute real path to prevent symlink/traversal attacks ---
    try:
        abs_path = os.path.realpath(os.path.abspath(file_path))
    except (OSError, ValueError) as exc:
        return {"status": "error", "reason": "Path resolution failed: {}".format(exc)}

    # --- Basic existence checks ---
    if not os.path.exists(abs_path):
        return {"status": "error", "reason": "File not found: {}".format(abs_path)}

    if not os.path.isfile(abs_path):
        return {"status": "error", "reason": "Path is not a regular file: {}".format(abs_path)}

    # --- Size cap (anti-DoS: reject zip bombs and excessively large files) ---
    try:
        file_size = os.path.getsize(abs_path)
    except OSError as exc:
        return {"status": "error", "reason": "Cannot determine file size: {}".format(exc)}

    if file_size > MAX_FILE_SIZE_BYTES:
        max_mb = MAX_FILE_SIZE_BYTES // (1024 * 1024)
        actual_mb = file_size / (1024 * 1024)
        return {
            "status": "error",
            "reason": "File too large: {:.1f} MB exceeds maximum allowed size of {} MB".format(
                actual_mb, max_mb
            ),
        }

    if file_size == 0:
        return {"status": "error", "reason": "File is empty (0 bytes) — nothing to analyze"}

    # --- Read permission check ---
    if not os.access(abs_path, os.R_OK):
        return {"status": "error", "reason": "Permission denied: cannot read file"}

    # --- Path traversal prevention ---
    # Convert to normalized absolute path and verify no directory escapes
    norm = os.path.normpath(abs_path)
    if ".." in norm.split(os.sep):
        return {"status": "error", "reason": "Path traversal detected in file path"}

    return None
